### PR TITLE
Feature/attachment upload athene

### PR DESCRIPTION
**Pull request by Adem-Khachnaoui**
### Checklist
- [ ] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [ ] Server: I added multiple integration tests (Spring) related to the features
- [ ] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [ ] Server: I implemented the changes with a good performance and prevented too many database calls
- [ ] Server: I documented the Java code using JavaDoc style.
- [ ] Client: I added multiple integration tests (Jest) related to the features
- [ ] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [ ] Client: I documented the TypeScript code using JSDoc style.
- [ ] Client: I added multiple screenshots/screencasts of my UI changes
- [ ] Client: I translated all the newly inserted strings into German and English

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Choose any lecture and add an attachment to it.
4. Check if a request is sent to localhost:8001/upload with the respective file/

### CODE REVIEW

**Comment by Adem-Khachnaoui** [comment]
Lecture attachments are now uploaded to Athene to be used as training material for the ELMo model of the respective course.

**Comment by krusche** [comment]
I think we should first discuss if uploading lecture slides to other systems is ok for instructors. As we now have more courses from other departments, they might not be happy if we upload and process their files without permission. Also, not all the attachments in a lecture include slides.

So we should have at least a checkbox in the attachment upload that the instructors needs to tick to approve the upload. Only then, the PDF should be uploaded to other systems

**Comment by Adem-Khachnaoui** [comment]
> I think we should first discuss if uploading lecture slides to other systems is ok for instructors. As we now have more courses from other departments, they might not be happy if we upload and process their files without permission. Also, not all the attachments in a lecture include slides.
> 
> So we should have at least a checkbox in the attachment upload that the instructors needs to tick to approve the upload. Only then, the PDF should be uploaded to other systems

@krusche 

**"I think we should first discuss if uploading lecture slides to other systems is ok for instructors"**
Yes, I discussed this with @jpbernius and we thought that Athene is somehow also a part of the Artemis system.

**"So we should have at least a checkbox in the attachment upload that the instructors needs to tick to approve the upload".**
yes that's definitely something I should add. 

**"Also, not all the attachments in a lecture include slides"**
For now we can extract training data from all pdfs and text files, But I will for now add a condition to only upload pdf files.

**Comment by krusche** [comment]
I guess if we have a checkbox and you only upload files if
1) The instructor has ticked the checkbox
2) The file is a PDF file

this would be fine!

**Comment by jpbernius** [comment]
> * The instructor has ticked the checkbox

Would we define this (a) once for the course or (b) for every attachment?

**Comment by krusche** [comment]
> > * The instructor has ticked the checkbox
> 
> Would we define this (a) once for the course or (b) for every attachment?

I would say for each attachment

**Comment by Adem-Khachnaoui** [comment]
> > * The instructor has ticked the checkbox
> 
> Would we define this (a) once for the course or (b) for every attachment?


Yes, I would also say for each attachment, as this enables the instructor to choose not to add attachments that are somehow out of the general context of the lecture (I'm thinking about attachments like event advertisements, Ferienakademie flyers, .... 😛   )

**Comment by Adem-Khachnaoui** [comment]
Capture

**Comment by Adem-Khachnaoui** [comment]
> I'm not sure if it happens in reality but I managed to get:
> 
> 1. `Exception in de.tum.in.www1.artemis.service.connectors.EmbeddingTrainingMaterialService.uploadAttachment() with cause = 'NULL' and exception = '404 Not Found: [no body]'`
>    by not building the right branch in Athene.
> 2. Exception in `de.tum.in.www1.artemis.service.connectors.EmbeddingTrainingMaterialService.uploadAttachment() with cause = 'NULL' and exception = '500 Internal Server Error: `
>    by giving wrong credentials to cloud.
> 
> The problem is I didn't get any message that my attachment is not saved to Athene.
> And deleting an attachment does not delete it from the cloud I suppose. (idk if it is intended)
> 
> Except from these, it worked as it is described :)
> 
> Edit: Maybe you can also say in your checkbox text that it only works for pdf files because I thought I could send a png too.


The first exception is expected if you don't run the Athene component.
Second exception should is dealt with in Athene now.

**" I didn't get any message that my attachment is not saved to Athene."**
you get  message by successful upload and the trace of the error otherwise.

**deleting an attachment does not delete it from the cloud I suppose**
@jpbernius : what do you think ? should we implement delete requests as well ?

**Comment by jpbernius** [comment]
> should we implement delete requests as well ?

I would consider this for a flow up but keep it as is for now

**Comment by bigultekin** [comment]
> > I'm not sure if it happens in reality but I managed to get:
> > 
> > 1. `Exception in de.tum.in.www1.artemis.service.connectors.EmbeddingTrainingMaterialService.uploadAttachment() with cause = 'NULL' and exception = '404 Not Found: [no body]'`
> >    by not building the right branch in Athene.
> > 2. Exception in `de.tum.in.www1.artemis.service.connectors.EmbeddingTrainingMaterialService.uploadAttachment() with cause = 'NULL' and exception = '500 Internal Server Error: `
> >    by giving wrong credentials to cloud.
> > 
> > The problem is I didn't get any message that my attachment is not saved to Athene.
> > And deleting an attachment does not delete it from the cloud I suppose. (idk if it is intended)
> > Except from these, it worked as it is described :)
> > Edit: Maybe you can also say in your checkbox text that it only works for pdf files because I thought I could send a png too.
> 
> The first exception is expected if you don't run the Athene component.
> Second exception should is dealt with in Athene now.
> 
> **" I didn't get any message that my attachment is not saved to Athene."**
> you get message by successful upload and the trace of the error otherwise.
> 
> **deleting an attachment does not delete it from the cloud I suppose**
> @jpbernius : what do you think ? should we implement delete requests as well ?
> 
> **Maybe you can also say in your checkbox text that it only works for pdf files because I thought I could send a png too.**
> I will do that !
> 
> Thanks for the feedback !

**"Second exception should is dealt with in Athene now."**
Yes, I saw that 👌

**"I will do that !"**
It looks good, I really liked it 👍

**"you get message by successful upload and the trace of the error otherwise."**
I meant in the user interface for the user but maybe it is not in the scope of this PR (could be good to have in the future in terms of UX)

**Comment by linusmichel** [comment]
Works as expected for me. Just experienced an issue when deleting an attachment, that the dialog stays open even if you already hit the Delete-Button (see Screenshot):

Bildschirmfoto 2020-06-15 um 13 23 32

**Comment by tiffca** [comment]
I wasn't able to deploy this in test server 1 or 3 using the following release: https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=277152368

After deployment on both servers, I'm not able to access the URLs. 
Deploying the "develop" branch on top fixes the issue after some time. 
Are we supposed to deploy on local only? 

Example message on test server 3 after deploying. Neither server is accessible even after 10+ minutes. 
image

**Comment by Adem-Khachnaoui** [comment]
> I wasn't able to deploy this in test server 1 or 3 using the following release: https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=277152368
> 
> After deployment on both servers, I'm not able to access the URLs.
> Deploying the "develop" branch on top fixes the issue after some time.
> Are we supposed to deploy on local only?
> 
> Example message on test server 3 after deploying. Neither server is accessible even after 10+ minutes.
> image

Thanks for testing, 
It's because we need to deploy an up-to-date version of Athene + update config files in test servers

File: src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        public String remotePath;
+    }
+    // endregion
+
+    @Value("${artemis.automatic-text.material-upload-url}")
+    private String API_ENDPOINT;
+
+    @Value("${artemis.automatic-text.secret}")
+    private String API_SECRET;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java:52]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968860)

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/EmbeddingTrainingMaterialScheduleService.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+@Profile("automaticText")
+public class EmbeddingTrainingMaterialScheduleService {
+
+    private final Logger log = LoggerFactory.getLogger(EmbeddingTrainingMaterialScheduleService.class);
+
+    private final Map<Long, ScheduledFuture> scheduledClusteringTasks = new HashMap<>();
+
+    private final EmbeddingTrainingMaterialService embeddingTrainingMaterialService;
+
+    private final Environment env;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/EmbeddingTrainingMaterialScheduleService.java:34]: Codacy found an issue: [Avoid unused private fields such as 'env'.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968860)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    private FileService fileService = new FileService();
+
+    private RemoteArtemisServiceConnector<EmbeddingTrainingMaterialService.Request, EmbeddingTrainingMaterialService.Response> connector = new RemoteArtemisServiceConnector<>(log,
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java:56]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968860)

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/EmbeddingTrainingMaterialScheduleService.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import de.tum.in.www1.artemis.domain.Attachment;
+import de.tum.in.www1.artemis.exception.NetworkingError;
+import de.tum.in.www1.artemis.security.SecurityUtils;
+import de.tum.in.www1.artemis.service.connectors.EmbeddingTrainingMaterialService;
+
+@Service
+@Profile("automaticText")
+public class EmbeddingTrainingMaterialScheduleService {
+
+    private final Logger log = LoggerFactory.getLogger(EmbeddingTrainingMaterialScheduleService.class);
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/EmbeddingTrainingMaterialScheduleService.java:28]: Codacy found an issue: [Avoid unused private fields such as 'log'.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968860)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    // endregion
+
+    @Value("${artemis.automatic-text.material-upload-url}")
+    private String API_ENDPOINT;
+
+    @Value("${artemis.automatic-text.secret}")
+    private String API_SECRET;
+
+    private FileService fileService = new FileService();
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java:54]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968860)

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/EmbeddingTrainingMaterialScheduleService.java (hunk 2 above)
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/EmbeddingTrainingMaterialScheduleService.java:34]: Codacy found an issue: [Perhaps 'env' could be replaced by a local variable.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968860)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    private static class Response {
+
+        public String remotePath;
+    }
+    // endregion
+
+    @Value("${artemis.automatic-text.material-upload-url}")
+    private String API_ENDPOINT;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/EmbeddingTrainingMaterialService.java:49]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968860)

**Comment by krusche** [comment]
Closing this for now. The PR is totally outdated. Please first finish the PR properly and then reopen it or even create a new one.

**Pull request closed without merging**